op-supervisor/supervisor/backend/db/logs/db.go

```diff
@@ -39,9 +30,168 @@ type Metrics interface {
 	RecordDBSearchEntriesRead(count int64)
 }
 
+// logContext is a buffer on top of the DB,
+// where blocks and logs can be applied to.
 type logContext struct {
-	blockNum uint64
-	logIdx   uint32
+	// current entry index, including the contents of `out`
+	entryIndex entrydb.EntryIdx
+
+	// current block data
+	blockHash types.TruncatedHash
+	blockNum  uint64
+	timestamp uint64
+	doneBlock bool // when the block data has been fully appended to out
+
+	// current log event data
+	logIdx  uint32
+	logHash types.TruncatedHash
+	execMsg *types.ExecutingMessage
+	doneLog bool // when the log data has been fully appended to out
+
+	// last type of entry we added
+	lastType entrydb.EntryType
+
+	// buffer of entries not yet in the DB
+	out []entrydb.Entry
+}
+
+type EntryObj interface {
+	encode() entrydb.Entry
+}
+
+func (l *logContext) BlockHash() types.TruncatedHash {
+	if !l.doneBlock {
+		panic("logContext in invalid state, cannot retrieve block ID")
+	}
+	return l.blockHash
+}
+
+// TODO other getters, with checks if doneBlock and doneLog
+
+func (l *logContext) applyEntry(entry entrydb.Entry) {
+	if len(l.out) != 0 {
+		panic("can only apply without appending if the state is still empty")
+	}
+
+	// TODO decode entry
+
+	// TODO apply to state
+
+	l.lastType = entry.Type()
+	l.entryIndex += 1
+}
+
+func (l *logContext) appendEntry(obj EntryObj) {
+	entry := obj.encode()
+	l.out = append(l.out, entry)
+	l.lastType = entry.Type()
+	l.entryIndex += 1
+}
+
+func (l *logContext) infer() error {
+	if (l.entryIndex+1)%searchCheckpointFrequency == 0 {
+		l.appendEntry(newSearchCheckpoint(l.blockNum, l.logIdx, l.timestamp))
+		return nil
+	}
+	startLog := func() error {
+		if l.doneLog {
+			return io.EOF
+		}
+		evt := newInitiatingEvent(l.logHash, l.execMsg != nil)
+		l.appendEntry(evt)
+		if l.execMsg == nil {
+			l.doneLog = true
+		}
+		return nil
+	}
+	switch l.lastType {
+	case entrydb.TypeSearchCheckpoint:
+		l.appendEntry(newCanonicalHash(l.blockHash))
+		return nil
+	case entrydb.TypeCanonicalHash:
+		l.doneBlock = true
+		return startLog()
```

Comment:
Just because a CanonicalHash was last appended does not mean the block is done, right? If a block had >256 log entries, it would emit a checkpoint and hash, but not be done.

Reply:
fixed the variable naming. The boolean was meant to indicate when we are done identifying the block.